presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionarySelectiveStreamReader.java

```diff
@@ -275,15 +275,12 @@ private void openRowGroup()
     {
         // read the dictionary
         if (!dictionaryOpen && dictionarySize > 0) {
-            if (dictionary.length < dictionarySize) {
-                dictionary = new long[dictionarySize];
-            }
-
-            LongInputStream dictionaryStream = dictionaryDataStreamSource.openStream();
-            if (dictionaryStream == null) {
-                throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Dictionary is not empty but data stream is not present");
-            }
-            dictionaryStream.nextLongVector(dictionarySize, dictionary);
+            GetDictionaryResult getDictionaryResult = dictionaryProvider.getDictionary(streamDescriptor, dictionary, dictionarySize);
+            dictionary = getDictionaryResult.getDictionaryBuffer();
+            // If using shared dictionary buffer, dictionaryMemoryUsage remains 0. Since flat map
+            // stream readers get reset in the next stripe, we don't need to distinguish carried
+            // over buffers and shared buffers.
+            dictionaryMemoryUsage = getDictionaryResult.getIsNewBuffer() ? sizeOf(dictionary) : dictionaryMemoryUsage;
```

Comment:
This is a weird way to computing memory usage for the dictionary. `dictionaryMemoryUsage` is not accumulative in this case, meaning that it is either `sizeOf(dictionary)` or `0`. So why don't we use `sizeOf(dictionary)` directly? Check my comment for `getRetainedSizeInBytes `

Reply:
I see, we need one state in the class to track whether we should account for the new dictionary size in `getRetainedSizeInBytes`. I chose to calculate dictionary memory usage directly in the method causing the memory change, while your preference was to use a boolean state and calculate it in a centralized place. 

I think the current way is nice in that it doesn't leak the concept of new buffer outside of the method, but I'm fine with the other way as well. Please let me know.